crates/katana/rpc/rpc/tests/starknet.rs

```diff
@@ -172,5 +176,171 @@ async fn test_send_declare_and_deploy_legacy_contract() {
+            .unwrap();
+    let starknet_account = katana_runner.account(0);
+
+    let path: PathBuf = PathBuf::from("tests/test_data/cairo_l1_msg_contract.json");
+    let (contract, compiled_class_hash) =
+        common::prepare_contract_declaration_params(&path).unwrap();
+
+    let class_hash = contract.class_hash();
+    let res =
+        starknet_account.declare(Arc::new(contract), compiled_class_hash).send().await.unwrap();
+
+    let receipt = TransactionWaiter::new(res.transaction_hash, starknet_account.provider())
+        .with_tx_status(TransactionFinalityStatus::AcceptedOnL2)
+        .await
+        .expect("Invalid tx receipt");
+
+    assert_eq!(receipt.finality_status(), &TransactionFinalityStatus::AcceptedOnL2);
+
+    assert!(
+        starknet_account
+            .provider()
+            .get_class(BlockId::Tag(BlockTag::Latest), class_hash)
+            .await
+            .is_ok()
+    );
+
+    let constructor_calldata = vec![];
+
+    let calldata = [
+        vec![
+            class_hash,                                     // class hash
+            FieldElement::ZERO,                             // salt
+            FieldElement::ZERO,                             // unique
+            FieldElement::from(constructor_calldata.len()), // constructor calldata len
+        ],
+        constructor_calldata.clone(),
+    ]
+    .concat();
+
+    let contract_address = get_contract_address(
+        FieldElement::ZERO,
+        res.class_hash,
+        &constructor_calldata,
+        FieldElement::ZERO,
+    );
+
+    let transaction = starknet_account
+        .execute(vec![Call {
+            calldata,
+            // devnet UDC address
+            to: FieldElement::from_hex_be(
+                "0x41a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf",
+            )
+            .unwrap(),
+            selector: get_selector_from_name("deployContract").unwrap(),
+        }])
+        .send()
+        .await
+        .unwrap();
+
+    // wait for the tx to be mined
+    TransactionWaiter::new(transaction.transaction_hash, starknet_account.provider())
+        .with_tx_status(TransactionFinalityStatus::AcceptedOnL2)
+        .await
+        .expect("Invalid tx receipt");
+
+    assert_eq!(
+        contract_address,
+        FieldElement::from_str(
+            "0x033d18fcfd3ae75ae4e8a275ce649220ed718b68dc53425b388fedcdbeab5097"
+        )
+        .unwrap()
+    );
+
+    let builder = contract_c1
+        .sendMessage(
+            U256::from_str("0x033d18fcfd3ae75ae4e8a275ce649220ed718b68dc53425b388fedcdbeab5097")
+                .unwrap(),
+            U256::from_str("0x005421de947699472df434466845d68528f221a52fce7ad2934c5dae2e1f1cdc")
+                .unwrap(),
+            vec![U256::from(123)],
+        )
+        .gas(12000000)
+        .value(Uint::from(1));
+
+    // Messaging between L1 -> L2
+
+    let _receipt = builder
```

Comment:
afaik the receipt type has a `status` field to check if the tx is successful or not. lets assert for that.